### PR TITLE
Add CONCATENATION to tflite frontend, support Inception V3

### PR DESCRIPTION
**Pull request by ariwaranosai**
add concatenation to tflite frontend to support inception v3.

### CODE REVIEW

**Comment by tqchen** [comment]
@kazum @FrozenGene @Huyuwei  please help review this PR

File: python/tvm/relay/frontend/tflite.py [hunk 1]
```diff
@@ -247,6 +249,55 @@ def convert_softmax(self, op):
...
+        except ImportError:
+            raise ImportError("The tflite package must be installed")
+
+        assert isinstance(op, Operator)
+        input_tensors = self.get_input_tensors(op)
+        output_tensors = self.get_output_tensors(op)
+        assert len(input_tensors) >= 1, "input tensors should greater than 1"
+        assert len(output_tensors) == 1, "output tensors should be 1"
+        in_exprs = [self.get_expr(input_tensor.tensor_idx) for input_tensor in input_tensors]
+        assert op.BuiltinOptionsType() == BuiltinOptions.ConcatenationOptions
```
Comment (FrozenGene) [inline python/tvm/relay/frontend/tflite.py]: Suggest adding additional one blank line to separate input / output logic and Options like we did in other implementations.

File: python/tvm/relay/frontend/tflite.py [hunk 2]
```diff
@@ -247,6 +249,55 @@ def convert_softmax(self, op):
...
+        op_options = op.BuiltinOptions()
+        concatenation_options = ConcatenationOptions()
+        concatenation_options.Init(op_options.Bytes, op_options.Pos)
+        concatenation_axis = concatenation_options.Axis()
+        fused_activation_fn = concatenation_options.FusedActivationFunction()
+        input_shape_length = len(input_tensors[0].tensor.ShapeAsNumpy())
+        axis_convert_map = {}
```
Comment (FrozenGene) [inline python/tvm/relay/frontend/tflite.py]: Suggest adding additional one blank line to indicate Options part is end like we did in other implementations.

File: python/tvm/relay/frontend/tflite.py [hunk 3]
```diff
@@ -247,6 +249,55 @@ def convert_softmax(self, op):
...
+        # TFLite is N H W C, our layout is N C H W
+        if input_shape_length == 1 or input_shape_length == 2:
+            # The rule is channel first (after N but before H, W).
+            # length of 1 means N*H*W*C, do nothing.
+            # length of 2 means N*H*W, C, do nothing.
+            pass
+        elif input_shape_length == 3:
+            # convert N C H*W to N H*W C
+            axis_convert_map = {1:2, 2:1}
```
Comment (FrozenGene) [inline python/tvm/relay/frontend/tflite.py]: line 284 and line 287. One has space between `:`,  one haven't. Please unify the style.

Additionally, your way of axis_convert_map could avoid additional transpose symbol. Could you help to improve `convert_reshape` and `convert_squeeze` too?

File: tests/python/frontend/tflite/test_forward.py [hunk 4]
```diff
@@ -355,11 +402,30 @@ def test_forward_mobilenet():
     tvm.testing.assert_allclose(np.squeeze(tvm_output[0]), np.squeeze(tflite_output[0]),
                                 rtol=1e-5, atol=1e-5)
 
+#######################################################################
+# Inception V3
+# ---------
```
Comment (srkreddy1238) [inline tests/python/frontend/tflite/test_forward.py]: Underline up to the heading pls.

File: tests/python/frontend/tflite/test_forward.py [hunk 5]
```diff
@@ -355,11 +402,30 @@ def test_forward_mobilenet():
...
+def test_forward_inception_v3_net():
+    '''test inception v3 tflite model'''
+    # MobilenetV1
```
Comment (srkreddy1238) [inline tests/python/frontend/tflite/test_forward.py]: MobilenetV1 -> InceptionV3

File: tests/python/frontend/tflite/test_forward.py [hunk 6]
```diff
@@ -355,11 +402,30 @@ def test_forward_mobilenet():
...
+    tflite_model_file = tf_testing.get_workload_official(
+        "https://storage.googleapis.com/download.tensorflow.org/models/tflite/model_zoo/upload_20180427/inception_v3_2018_04_27.tgz",
+        "inception_v3.tflite", temp)
+    tflite_model_buf = open(tflite_model_file, "rb").read()
+    data = np.random.uniform(size=(1, 299, 299, 3)).astype('float32')
+    tvm_data = np.transpose(data, axes=(0, 3, 1, 2))
+    tflite_output = run_tflite_graph(tflite_model_buf, data)
+    tvm_output = run_tvm_graph(tflite_model_buf, tvm_data, 'input')
+    tvm.testing.assert_allclose(np.squeeze(tvm_output[0]), np.squeeze(tflite_output[0]),
+                                rtol=1e-5, atol=1e-5)
```
Comment (srkreddy1238) [inline tests/python/frontend/tflite/test_forward.py]: Close the tflite file and remove the tmpdir

File: python/tvm/relay/frontend/tflite.py (hunk 3 above)
Comment (ariwaranosai) [reply python/tvm/relay/frontend/tflite.py]: It's different between concat and reshape. To remove additional op, we have to  use code likes [code in tfcoreml](https://github.com/tf-coreml/tf-coreml/blob/1e9e90bad799db7eb8d5547bdbeb0cc1e5180007/tfcoreml/_shape_sensitive_layers.py#L282). I'll try to change those later with support models as much as possible.

**Comment by ariwaranosai** [comment]
@FrozenGene @srkreddy1238 fixed, please check again.

File: python/tvm/relay/frontend/tflite.py [hunk 7]
```diff
@@ -247,6 +249,58 @@ def convert_softmax(self, op):
...
+
+        axis_convert_map = {}
+        # TFLite is N H W C, our layout is N C H W
+        if input_shape_length == 1 or input_shape_length == 2:
+            # The rule is channel first (after N but before H, W).
+            # length of 1 means N*H*W*C, do nothing.
+            # length of 2 means N*H*W, C, do nothing.
+            pass
+        elif input_shape_length == 3:
+            # convert N C H*W to N H*W C
```
Comment (kazum) [inline python/tvm/relay/frontend/tflite.py]: convert N H&ast;W C to N C H&ast;W?

File: python/tvm/relay/frontend/tflite.py [hunk 8]
```diff
@@ -247,6 +249,58 @@ def convert_softmax(self, op):
...
+            axis_convert_map = {1: 2, 2: 1}
+        elif input_shape_length == 4:
+            # change axis.
+            axis_convert_map = {1: 2, 2: 3, 3: 1}
+        else:
+            raise NotImplementedError("Not support input shape length {} of concatenatio : "
+                                      .format(str(input_shape_length)))
+
+        # axis in N H W C
+        concatenation_axis = axis_convert_map.get(concatenation_axis, concatenation_axis)
```
Comment (kazum) [inline python/tvm/relay/frontend/tflite.py]: I guess we can write more generally and simply.
```
# convert map from channels_last to channels_first layout
axis_convert_map = [0] + list(range(2, input_shape_length)) + [1]
concatenation_axis = axis_convert_map[concatenation_axis]
```

Comment (ariwaranosai) [reply python/tvm/relay/frontend/tflite.py]: good idea, fixed.

**Comment by tqchen** [comment]
@srkreddy1238 please follow up on this PR

**Comment by tqchen** [comment]
Thanks @kazum @FrozenGene for reviewing

**Pull request merged**